# Segmenting and Clustering Neighbourhoods in Toronto
[Week 3 Assignment Segmenting and Clustering Neighbourhoods in Toronto](https://github.com/ChrisFord2006/ADS_Capstone_Week1)

In [1]:
pip install geopy

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
Note: you may need to restart the kernel to use updated packages.


In [2]:
pip install folium

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/dhcrypto.py:16: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/secretstorage/util.py:25: CryptographyDeprecationWarning: int_from_bytes is deprecated, use int.from_bytes instead
  from cryptography.utils import int_from_bytes
     |████████████████████████████████| 94 kB 7.0 MB/s  eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [3]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [4]:
html_data = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text
#html_data

In [5]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(html_data, 'html.parser')

In [6]:
soup

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"3ade1005-914c-482b-85fd-9a2ba97a08f6","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":1029579868,"wgRevisionId":1029579868,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Short description is different from Wikidata","Communications in Ontario"

### Processing the data into a Pandas Data Frame

In [7]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [8]:
df.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A      Queen's Park   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                         Neighborhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4                       Ontario Provincial Government  
5                                    Islington Avenue  
6                                      Malvern, Rouge  
7                                     Don Mills North  
8                     Parkview Hill, Woodbine Gardens  
9                            Garden District, Ryerson  
10                                          Glencairn  
11  West Deane Park, Princess Gardens, Martin Grov...

In [9]:
df.shape

(103, 3)

### Adding in geospatial data to the above dataframe
#### First great data frame for geospatial data
#### Then sort postcodes in both dataframes so they match up correctly when latitude and longitude columns added.
[Loading and accessing data in a Notebook Watson Studio](https://www.ibm.com/support/producthub/icpdata/docs/content/SSQNUZ_latest/wsj/analyze-data/load-and-access-data.html)

#### df_data_1 = pd.read_csv('Geospatial_Coordinates.csv') this line of code doesn't work in the IBM Cloud site the code below does.
#### df_data_1 = pd.read_csv('Geospatial_Coordinates.csv') does work in the IBM Skills Network lab with ('Geospatial_Coordinates.csv') in the same folder as the notebook

In [10]:
#
import os, types
import pandas as pd
from botocore.client import Config
import ibm_boto3

def __iter__(self): return 0

# @hidden_cell
# The following code accesses a file in your IBM Cloud Object Storage. It includes your credentials.
# You might want to remove those credentials before you share the notebook.

if os.environ.get('RUNTIME_ENV_LOCATION_TYPE') == 'external':
    endpoint_23e3b5aeac9043948fc832b12320c85a = 'https://s3.us.cloud-object-storage.appdomain.cloud'
else:
    endpoint_23e3b5aeac9043948fc832b12320c85a = 'https://s3.private.us.cloud-object-storage.appdomain.cloud'

client_23e3b5aeac9043948fc832b12320c85a = ibm_boto3.client(service_name='s3',
    ibm_api_key_id='',
    ibm_auth_endpoint="https://iam.cloud.ibm.com/oidc/token",
    config=Config(signature_version='oauth'),
    endpoint_url=endpoint_23e3b5aeac9043948fc832b12320c85a)

body = client.get_object(Bucket='ibmdscapstoneweel3toronto-donotdelete-pr-uew1iz4itwgorw',Key='Geospatial_Coordinates.csv')['Body']
# add missing __iter__ method, so pandas accepts body as file-like object
if not hasattr(body, "__iter__"): body.__iter__ = types.MethodType( __iter__, body )

df_data_1 = pd.read_csv(body)
df_data_1.head(10)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476
5         M1J  43.744734 -79.239476
6         M1K  43.727929 -79.262029
7         M1L  43.711112 -79.284577
8         M1M  43.716316 -79.239476
9         M1N  43.692657 -79.264848

In [11]:
df_data_1.shape

(103, 3)

In [12]:
df1= df.sort_values('PostalCode')
df1A = df1.reset_index()
df1A

index PostalCode                 Borough  \
0        6        M1B             Scarborough   
1       12        M1C             Scarborough   
2       18        M1E             Scarborough   
3       22        M1G             Scarborough   
4       26        M1H             Scarborough   
5       32        M1J             Scarborough   
6       38        M1K             Scarborough   
7       44        M1L             Scarborough   
8       51        M1M             Scarborough   
9       58        M1N             Scarborough   
10      65        M1P             Scarborough   
11      71        M1R             Scarborough   
12      78        M1S             Scarborough   
13      82        M1T             Scarborough   
14      85        M1V             Scarborough   
15      90        M1W             Scarborough   
16      95        M1X             Scarborough   
17      27        M2H              North York   
18      33        M2J              North York   
19      39        M2K              North York   
20      45        M2L              North York   
21      52        M2M              North York   
22      59        M2N              North York   
23      66        M2P              North York   
24      72        M2R              North York   
25       0        M3A              North York   
26       7        M3B              North York   
27      13        M3C              North York   
28      28        M3H              North York   
29      34        M3J              North York   
30      40        M3K              North York   
31      46        M3L              North York   
32      53        M3M              North York   
33      60        M3N              North York   
34       1        M4A              North York   
35       8        M4B               East York   
36      14        M4C               East York   
37      19        M4E            East Toronto   
38      23        M4G               East York   
39      29        M4H               East York   
40      35        M4J  East York/East Toronto   
41      41        M4K            East Toronto   
42      47        M4L            East Toronto   
43      54        M4M            East Toronto   
44      61        M4N         Central Toronto   
45      67        M4P         Central Toronto   
46      73        M4R         Central Toronto   
47      79        M4S         Central Toronto   
48      83        M4T         Central Toronto   
49      86        M4V         Central Toronto   
50      91        M4W        Downtown Toronto   
51      96        M4X        Downtown Toronto   
52      99        M4Y        Downtown Toronto   
53       2        M5A        Downtown Toronto   
54       9        M5B        Downtown Toronto   
55      15        M5C        Downtown Toronto   
56      20        M5E        Downtown Toronto   
57      24        M5G        Downtown Toronto   
58      30        M5H        Downtown Toronto   
59      36        M5J        Downtown Toronto   
60      42        M5K        Downtown Toronto   
61      48        M5L        Downtown Toronto   
62      55        M5M              North York   
63      62        M5N         Central Toronto   
64      68        M5P         Central Toronto   
65      74        M5R         Central Toronto   
66      80        M5S        Downtown Toronto   
67      84        M5T        Downtown Toronto   
68      87        M5V        Downtown Toronto   
69      92        M5W  Downtown Toronto Stn A   
70      97        M5X        Downtown Toronto   
71       3        M6A              North York   
72      10        M6B              North York   
73      16        M6C                    York   
74      21        M6E                    York   
75      25        M6G        Downtown Toronto   
76      31        M6H            West Toronto   
77      37        M6J            West Toronto   
78      43        M6K            West Toronto   
79      49        M6L              North York   
80      56        M6M               

In [13]:
df1B = df1A.drop(columns=['index'])
df1B

PostalCode                 Borough  \
0          M1B             Scarborough   
1          M1C             Scarborough   
2          M1E             Scarborough   
3          M1G             Scarborough   
4          M1H             Scarborough   
5          M1J             Scarborough   
6          M1K             Scarborough   
7          M1L             Scarborough   
8          M1M             Scarborough   
9          M1N             Scarborough   
10         M1P             Scarborough   
11         M1R             Scarborough   
12         M1S             Scarborough   
13         M1T             Scarborough   
14         M1V             Scarborough   
15         M1W             Scarborough   
16         M1X             Scarborough   
17         M2H              North York   
18         M2J              North York   
19         M2K              North York   
20         M2L              North York   
21         M2M              North York   
22         M2N              North York   
23         M2P              North York   
24         M2R              North York   
25         M3A              North York   
26         M3B              North York   
27         M3C              North York   
28         M3H              North York   
29         M3J              North York   
30         M3K              North York   
31         M3L              North York   
32         M3M              North York   
33         M3N              North York   
34         M4A              North York   
35         M4B               East York   
36         M4C               East York   
37         M4E            East Toronto   
38         M4G               East York   
39         M4H               East York   
40         M4J  East York/East Toronto   
41         M4K            East Toronto   
42         M4L            East Toronto   
43         M4M            East Toronto   
44         M4N         Central Toronto   
45         M4P         Central Toronto   
46         M4R         Central Toronto   
47         M4S         Central Toronto   
48         M4T         Central Toronto   
49         M4V         Central Toronto   
50         M4W        Downtown Toronto   
51         M4X        Downtown Toronto   
52         M4Y        Downtown Toronto   
53         M5A        Downtown Toronto   
54         M5B        Downtown Toronto   
55         M5C        Downtown Toronto   
56         M5E        Downtown Toronto   
57         M5G        Downtown Toronto   
58         M5H        Downtown Toronto   
59         M5J        Downtown Toronto   
60         M5K        Downtown Toronto   
61         M5L        Downtown Toronto   
62         M5M              North York   
63         M5N         Central Toronto   
64         M5P         Central Toronto   
65         M5R         Central Toronto   
66         M5S        Downtown Toronto   
67         M5T        Downtown Toronto   
68         M5V        Downtown Toronto   
69         M5W  Downtown Toronto Stn A   
70         M5X        Downtown Toronto   
71         M6A              North York   
72         M6B              North York   
73         M6C                    York   
74         M6E                    York   
75         M6G        Downtown Toronto   
76         M6H            West Toronto   
77         M6J            West Toronto   
78         M6K            West Toronto   
79         M6L              North York   
80         M6M                    York   
81         M6N                    York   
82         M6P            West Toronto   
83         M6R            West Toronto   
84         M6S            West Toronto   
85         M7A            Queen's Park   
86         M7R             Mississauga   
87         M7Y   East Toronto Business   
88         M8V               Etobicoke   
89         M8W               Etobicoke   
90         M8X               Etobicoke   
91         M8Y               Etobicoke   
92         M8Z               Etobicoke   
93         M9A               Etobicoke   
94         M9B

In [14]:
df2 = pd.concat([df1B,df_data_1], axis=1)
df2

PostalCode                 Borough  \
0          M1B             Scarborough   
1          M1C             Scarborough   
2          M1E             Scarborough   
3          M1G             Scarborough   
4          M1H             Scarborough   
5          M1J             Scarborough   
6          M1K             Scarborough   
7          M1L             Scarborough   
8          M1M             Scarborough   
9          M1N             Scarborough   
10         M1P             Scarborough   
11         M1R             Scarborough   
12         M1S             Scarborough   
13         M1T             Scarborough   
14         M1V             Scarborough   
15         M1W             Scarborough   
16         M1X             Scarborough   
17         M2H              North York   
18         M2J              North York   
19         M2K              North York   
20         M2L              North York   
21         M2M              North York   
22         M2N              North York   
23         M2P              North York   
24         M2R              North York   
25         M3A              North York   
26         M3B              North York   
27         M3C              North York   
28         M3H              North York   
29         M3J              North York   
30         M3K              North York   
31         M3L              North York   
32         M3M              North York   
33         M3N              North York   
34         M4A              North York   
35         M4B               East York   
36         M4C               East York   
37         M4E            East Toronto   
38         M4G               East York   
39         M4H               East York   
40         M4J  East York/East Toronto   
41         M4K            East Toronto   
42         M4L            East Toronto   
43         M4M            East Toronto   
44         M4N         Central Toronto   
45         M4P         Central Toronto   
46         M4R         Central Toronto   
47         M4S         Central Toronto   
48         M4T         Central Toronto   
49         M4V         Central Toronto   
50         M4W        Downtown Toronto   
51         M4X        Downtown Toronto   
52         M4Y        Downtown Toronto   
53         M5A        Downtown Toronto   
54         M5B        Downtown Toronto   
55         M5C        Downtown Toronto   
56         M5E        Downtown Toronto   
57         M5G        Downtown Toronto   
58         M5H        Downtown Toronto   
59         M5J        Downtown Toronto   
60         M5K        Downtown Toronto   
61         M5L        Downtown Toronto   
62         M5M              North York   
63         M5N         Central Toronto   
64         M5P         Central Toronto   
65         M5R         Central Toronto   
66         M5S        Downtown Toronto   
67         M5T        Downtown Toronto   
68         M5V        Downtown Toronto   
69         M5W  Downtown Toronto Stn A   
70         M5X        Downtown Toronto   
71         M6A              North York   
72         M6B              North York   
73         M6C                    York   
74         M6E                    York   
75         M6G        Downtown Toronto   
76         M6H            West Toronto   
77         M6J            West Toronto   
78         M6K            West Toronto   
79         M6L              North York   
80         M6M                    York   
81         M6N                    York   
82         M6P            West Toronto   
83         M6R            West Toronto   
84         M6S            West Toronto   
85         M7A            Queen's Park   
86         M7R             Mississauga   
87         M7Y   East Toronto Business   
88         M8V               Etobicoke   
89         M8W               Etobicoke   
90         M8X               Etobicoke   
91         M8Y               Etobicoke   
92         M8Z               Etobicoke   
93         M9A               Etobicoke   
94         M9B

In [15]:
df3 = df2.drop(columns=['Postal Code'])
df3

PostalCode                 Borough  \
0          M1B             Scarborough   
1          M1C             Scarborough   
2          M1E             Scarborough   
3          M1G             Scarborough   
4          M1H             Scarborough   
5          M1J             Scarborough   
6          M1K             Scarborough   
7          M1L             Scarborough   
8          M1M             Scarborough   
9          M1N             Scarborough   
10         M1P             Scarborough   
11         M1R             Scarborough   
12         M1S             Scarborough   
13         M1T             Scarborough   
14         M1V             Scarborough   
15         M1W             Scarborough   
16         M1X             Scarborough   
17         M2H              North York   
18         M2J              North York   
19         M2K              North York   
20         M2L              North York   
21         M2M              North York   
22         M2N              North York   
23         M2P              North York   
24         M2R              North York   
25         M3A              North York   
26         M3B              North York   
27         M3C              North York   
28         M3H              North York   
29         M3J              North York   
30         M3K              North York   
31         M3L              North York   
32         M3M              North York   
33         M3N              North York   
34         M4A              North York   
35         M4B               East York   
36         M4C               East York   
37         M4E            East Toronto   
38         M4G               East York   
39         M4H               East York   
40         M4J  East York/East Toronto   
41         M4K            East Toronto   
42         M4L            East Toronto   
43         M4M            East Toronto   
44         M4N         Central Toronto   
45         M4P         Central Toronto   
46         M4R         Central Toronto   
47         M4S         Central Toronto   
48         M4T         Central Toronto   
49         M4V         Central Toronto   
50         M4W        Downtown Toronto   
51         M4X        Downtown Toronto   
52         M4Y        Downtown Toronto   
53         M5A        Downtown Toronto   
54         M5B        Downtown Toronto   
55         M5C        Downtown Toronto   
56         M5E        Downtown Toronto   
57         M5G        Downtown Toronto   
58         M5H        Downtown Toronto   
59         M5J        Downtown Toronto   
60         M5K        Downtown Toronto   
61         M5L        Downtown Toronto   
62         M5M              North York   
63         M5N         Central Toronto   
64         M5P         Central Toronto   
65         M5R         Central Toronto   
66         M5S        Downtown Toronto   
67         M5T        Downtown Toronto   
68         M5V        Downtown Toronto   
69         M5W  Downtown Toronto Stn A   
70         M5X        Downtown Toronto   
71         M6A              North York   
72         M6B              North York   
73         M6C                    York   
74         M6E                    York   
75         M6G        Downtown Toronto   
76         M6H            West Toronto   
77         M6J            West Toronto   
78         M6K            West Toronto   
79         M6L              North York   
80         M6M                    York   
81         M6N                    York   
82         M6P            West Toronto   
83         M6R            West Toronto   
84         M6S            West Toronto   
85         M7A            Queen's Park   
86         M7R             Mississauga   
87         M7Y   East Toronto Business   
88         M8V               Etobicoke   
89         M8W               Etobicoke   
90         M8X               Etobicoke   
91         M8Y               Etobicoke   
92         M8Z               Etobicoke   
93         M9A               Etobicoke   
94         M9B

Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you. 

Just make sure:

to add enough Markdown cells to explain what you decided to do and to report any observations you make. 
to generate maps to visualize your neighborhoods and how they cluster together. 

In [17]:
df5= df3.sort_values('Borough')

In [18]:
df5

PostalCode                 Borough  \
47         M4S         Central Toronto   
63         M5N         Central Toronto   
46         M4R         Central Toronto   
64         M5P         Central Toronto   
65         M5R         Central Toronto   
48         M4T         Central Toronto   
44         M4N         Central Toronto   
45         M4P         Central Toronto   
49         M4V         Central Toronto   
50         M4W        Downtown Toronto   
52         M4Y        Downtown Toronto   
53         M5A        Downtown Toronto   
54         M5B        Downtown Toronto   
56         M5E        Downtown Toronto   
57         M5G        Downtown Toronto   
58         M5H        Downtown Toronto   
60         M5K        Downtown Toronto   
61         M5L        Downtown Toronto   
55         M5C        Downtown Toronto   
51         M4X        Downtown Toronto   
67         M5T        Downtown Toronto   
66         M5S        Downtown Toronto   
75         M6G        Downtown Toronto   
59         M5J        Downtown Toronto   
68         M5V        Downtown Toronto   
70         M5X        Downtown Toronto   
69         M5W  Downtown Toronto Stn A   
43         M4M            East Toronto   
42         M4L            East Toronto   
41         M4K            East Toronto   
37         M4E            East Toronto   
87         M7Y   East Toronto Business   
39         M4H               East York   
38         M4G               East York   
36         M4C               East York   
35         M4B               East York   
40         M4J  East York/East Toronto   
93         M9A               Etobicoke   
101        M9V               Etobicoke   
92         M8Z               Etobicoke   
91         M8Y               Etobicoke   
90         M8X               Etobicoke   
89         M8W               Etobicoke   
88         M8V               Etobicoke   
99         M9P               Etobicoke   
100        M9R               Etobicoke   
95         M9C               Etobicoke   
94         M9B               Etobicoke   
102        M9W     Etobicoke Northwest   
86         M7R             Mississauga   
33         M3N              North York   
71         M6A              North York   
96         M9L              North York   
97         M9M              North York   
79         M6L              North York   
17         M2H              North York   
18         M2J              North York   
19         M2K              North York   
20         M2L              North York   
62         M5M              North York   
72         M6B              North York   
21         M2M              North York   
23         M2P              North York   
24         M2R              North York   
25         M3A              North York   
26         M3B              North York   
27         M3C              North York   
28         M3H              North York   
29         M3J              North York   
30         M3K              North York   
31         M3L              North York   
32         M3M              North York   
22         M2N              North York   
34         M4A              North York   
85         M7A            Queen's Park   
3          M1G             Scarborough   
5          M1J             Scarborough   
6          M1K             Scarborough   
7          M1L             Scarborough   
2          M1E             Scarborough   
8          M1M             Scarborough   
9          M1N             Scarborough   
10         M1P             Scarborough   
11         M1R             Scarborough   
4          M1H             Scarborough   
1          M1C             Scarborough   
12         M1S             Scarborough   
13         M1T             Scarborough   
14         M1V             Scarborough   
15         M1W             Scarborough   
16         M1X             Scarborough   
0          M1B             Scarborough   
83         M6R            West Toronto   
82         M6P            West Toronto   
78         M6K

## Creating a data frame with boroughs with the name toronto

In [19]:
df_Toronto = df5.iloc[0:32]
df_Toronto

PostalCode                 Borough  \
47        M4S         Central Toronto   
63        M5N         Central Toronto   
46        M4R         Central Toronto   
64        M5P         Central Toronto   
65        M5R         Central Toronto   
48        M4T         Central Toronto   
44        M4N         Central Toronto   
45        M4P         Central Toronto   
49        M4V         Central Toronto   
50        M4W        Downtown Toronto   
52        M4Y        Downtown Toronto   
53        M5A        Downtown Toronto   
54        M5B        Downtown Toronto   
56        M5E        Downtown Toronto   
57        M5G        Downtown Toronto   
58        M5H        Downtown Toronto   
60        M5K        Downtown Toronto   
61        M5L        Downtown Toronto   
55        M5C        Downtown Toronto   
51        M4X        Downtown Toronto   
67        M5T        Downtown Toronto   
66        M5S        Downtown Toronto   
75        M6G        Downtown Toronto   
59        M5J        Downtown Toronto   
68        M5V        Downtown Toronto   
70        M5X        Downtown Toronto   
69        M5W  Downtown Toronto Stn A   
43        M4M            East Toronto   
42        M4L            East Toronto   
41        M4K            East Toronto   
37        M4E            East Toronto   
87        M7Y   East Toronto Business   

                                         Neighborhood   Latitude  Longitude  
47                                         Davisville  43.704324 -79.388790  
63                                           Roselawn  43.711695 -79.416936  
46                                 North Toronto West  43.715383 -79.405678  
64                           Forest Hill North & West  43.696948 -79.411307  
65                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
55                                     St. James Town  43.651494 -79.375418  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
67          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
66                     University of Toronto, Harbord  43.662696 -79.400049  
75                                           Christie  43.669542 -79.422564  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
68  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
70             First Canadian Place, Underground city  43.648429 -79.382280  
69                                     Enclave of M5E  43.646435 -79.374846  
43                                    Studio District  43.659526 -79.340923  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
41                       The Danforth West, Riverdale  43.679557 -79.352188  
37                                        The Beaches  43.676357 -79.293031  
87                                     Enclave of M4L  43.662744 -79.321558

## Getting the Geographical Coordinates of Toronto City

In [20]:
address = 'Toronto City, ON'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [21]:
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_Toronto['Latitude'], df_Toronto['Longitude'], df_Toronto['Borough'], df_Toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [22]:
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df5['Latitude'], df5['Longitude'], df5['Borough'], df5['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [23]:
df_WToronto = df5.iloc[92:98]

In [24]:
df_WToronto

PostalCode       Borough                                  Neighborhood  \
83        M6R  West Toronto                        Parkdale, Roncesvalles   
82        M6P  West Toronto                 High Park, The Junction South   
78        M6K  West Toronto  Brockton, Parkdale Village, Exhibition Place   
77        M6J  West Toronto                      Little Portugal, Trinity   
76        M6H  West Toronto                  Dufferin, Dovercourt Village   
84        M6S  West Toronto                            Runnymede, Swansea   

     Latitude  Longitude  
83  43.648960 -79.456325  
82  43.661608 -79.464763  
78  43.636847 -79.428191  
77  43.647927 -79.419750  
76  43.669005 -79.442259  
84  43.651571 -79.484450

In [25]:
df_ATC = pd.concat([df_Toronto,df_WToronto])
df_ATC

PostalCode                 Borough  \
47        M4S         Central Toronto   
63        M5N         Central Toronto   
46        M4R         Central Toronto   
64        M5P         Central Toronto   
65        M5R         Central Toronto   
48        M4T         Central Toronto   
44        M4N         Central Toronto   
45        M4P         Central Toronto   
49        M4V         Central Toronto   
50        M4W        Downtown Toronto   
52        M4Y        Downtown Toronto   
53        M5A        Downtown Toronto   
54        M5B        Downtown Toronto   
56        M5E        Downtown Toronto   
57        M5G        Downtown Toronto   
58        M5H        Downtown Toronto   
60        M5K        Downtown Toronto   
61        M5L        Downtown Toronto   
55        M5C        Downtown Toronto   
51        M4X        Downtown Toronto   
67        M5T        Downtown Toronto   
66        M5S        Downtown Toronto   
75        M6G        Downtown Toronto   
59        M5J        Downtown Toronto   
68        M5V        Downtown Toronto   
70        M5X        Downtown Toronto   
69        M5W  Downtown Toronto Stn A   
43        M4M            East Toronto   
42        M4L            East Toronto   
41        M4K            East Toronto   
37        M4E            East Toronto   
87        M7Y   East Toronto Business   
83        M6R            West Toronto   
82        M6P            West Toronto   
78        M6K            West Toronto   
77        M6J            West Toronto   
76        M6H            West Toronto   
84        M6S            West Toronto   

                                         Neighborhood   Latitude  Longitude  
47                                         Davisville  43.704324 -79.388790  
63                                           Roselawn  43.711695 -79.416936  
46                                 North Toronto West  43.715383 -79.405678  
64                           Forest Hill North & West  43.696948 -79.411307  
65                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
55                                     St. James Town  43.651494 -79.375418  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
67          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
66                     University of Toronto, Harbord  43.662696 -79.400049  
75                                           Christie  43.669542 -79.422564  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
68  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
70             First Canadian Place, Underground city  43.648429 -79.382280  
69                                     Enclave of M5E  43.646435 -79.374846  
43                                    Studio District  43.659526 -79.340923  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
41                       The Danforth West, Riverdale  43.67955

In [26]:
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_ATC['Latitude'], df_ATC['Longitude'], df_ATC['Borough'], df_ATC['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [27]:
df_EY = df5.iloc[32:37]
df_EY

PostalCode                 Borough                     Neighborhood  \
39        M4H               East York                 Thorncliffe Park   
38        M4G               East York                          Leaside   
36        M4C               East York                 Woodbine Heights   
35        M4B               East York  Parkview Hill, Woodbine Gardens   
40        M4J  East York/East Toronto               The Danforth  East   

     Latitude  Longitude  
39  43.705369 -79.349372  
38  43.709060 -79.363452  
36  43.695344 -79.318389  
35  43.706397 -79.309937  
40  43.685347 -79.338106

In [28]:
df_TOR = pd.concat([df_ATC,df_EY])
df_TOR

PostalCode                 Borough  \
47        M4S         Central Toronto   
63        M5N         Central Toronto   
46        M4R         Central Toronto   
64        M5P         Central Toronto   
65        M5R         Central Toronto   
48        M4T         Central Toronto   
44        M4N         Central Toronto   
45        M4P         Central Toronto   
49        M4V         Central Toronto   
50        M4W        Downtown Toronto   
52        M4Y        Downtown Toronto   
53        M5A        Downtown Toronto   
54        M5B        Downtown Toronto   
56        M5E        Downtown Toronto   
57        M5G        Downtown Toronto   
58        M5H        Downtown Toronto   
60        M5K        Downtown Toronto   
61        M5L        Downtown Toronto   
55        M5C        Downtown Toronto   
51        M4X        Downtown Toronto   
67        M5T        Downtown Toronto   
66        M5S        Downtown Toronto   
75        M6G        Downtown Toronto   
59        M5J        Downtown Toronto   
68        M5V        Downtown Toronto   
70        M5X        Downtown Toronto   
69        M5W  Downtown Toronto Stn A   
43        M4M            East Toronto   
42        M4L            East Toronto   
41        M4K            East Toronto   
37        M4E            East Toronto   
87        M7Y   East Toronto Business   
83        M6R            West Toronto   
82        M6P            West Toronto   
78        M6K            West Toronto   
77        M6J            West Toronto   
76        M6H            West Toronto   
84        M6S            West Toronto   
39        M4H               East York   
38        M4G               East York   
36        M4C               East York   
35        M4B               East York   
40        M4J  East York/East Toronto   

                                         Neighborhood   Latitude  Longitude  
47                                         Davisville  43.704324 -79.388790  
63                                           Roselawn  43.711695 -79.416936  
46                                 North Toronto West  43.715383 -79.405678  
64                           Forest Hill North & West  43.696948 -79.411307  
65                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
55                                     St. James Town  43.651494 -79.375418  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
67          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
66                     University of Toronto, Harbord  43.662696 -79.400049  
75                                           Christie  43.669542 -79.422564  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
68  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
70             First Canadian Place, Underground city  43.648429 -79.382280  
69                                     Enclave of M5E  43.646435 -79.374846  
43            

In [29]:
# create map of Toronto using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_TOR['Latitude'], df_TOR['Longitude'], df_TOR['Borough'], df_TOR['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

## Foursquare Credentials

In [1]:
CLIENT_ID = '' # your Foursquare ID
CLIENT_SECRET = '' # your Foursquare Secret
ACCESS_TOKEN = '' # your FourSquare Acces
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 
CLIENT_SECRET:


## Explore Garden District, Ryerson, Downtown Toronto, LL Downtown Toronto Garden District, Ryerson	43.657162	-79.378937

In [31]:
df_TOR.loc[54, 'Neighborhood']

'Garden District, Ryerson'

In [32]:
neighborhood_latitude = df_TOR.loc[54, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_TOR.loc[54, 'Longitude'] # neighborhood longitude value

neighborhood_name = df_TOR.loc[54, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Garden District, Ryerson are 43.6571618, -79.37893709999999.


In [33]:

LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius


url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL



'https://api.foursquare.com/v2/venues/explore?&client_id=4ECWPTCW4UGYZURT4HN22B5NTHU2GRMI5HARQVLDCRBEOX42&client_secret=W0R3HIC4JEFJMNIHJCGPFC3NSPTUVSD43RCOKI0LIPQOEZYC&v=20180604&ll=43.6571618,-79.37893709999999&radius=500&limit=100'

In [34]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '60d6fda9081d00029c183b63'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Downtown Toronto',
  'headerFullLocation': 'Downtown Toronto, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 122,
  'suggestedBounds': {'ne': {'lat': 43.6616618045, 'lng': -79.37272880013877},
   'sw': {'lat': 43.652661795499995, 'lng': -79.3851453998612}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '56d4d1b3cd1035fe77e1492c',
       'name': 'Page One Cafe',
       'location': {'address': '106 Mutual St',
        'crossStreet': 'btwn Dundas & Gould St',
        'lat': 43.65777161112601,
        'lng': -79.3760725691681,
        'labeledLat

In [35]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [37]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  app.launch_new_instance()


name     categories        lat        lng
0        Page One Cafe           Café  43.657772 -79.376073
1          Blaze Pizza    Pizza Place  43.656518 -79.380015
2          Oakham Café           Café  43.658078 -79.378315
3  Silver Snail Comics     Comic Shop  43.657031 -79.381403
4         Burrito Boyz  Burrito Place  43.656265 -79.378343

In [38]:
nearby_venues.head(10)

name                 categories        lat  \
0                       Page One Cafe                       Café  43.657772   
1                         Blaze Pizza                Pizza Place  43.656518   
2                         Oakham Café                       Café  43.658078   
3                 Silver Snail Comics                 Comic Shop  43.657031   
4                        Burrito Boyz              Burrito Place  43.656265   
5                         UNIQLO ユニクロ             Clothing Store  43.655910   
6                         Milo's Pita         Falafel Restaurant  43.656270   
7                         Jazz Bistro                Music Venue  43.655678   
8  Ali Baba's - Yonge north of Dundas  Middle Eastern Restaurant  43.657867   
9                        Banh Mi Boys             Sandwich Place  43.659292   

         lng  
0 -79.376073  
1 -79.380015  
2 -79.378315  
3 -79.381403  
4 -79.378343  
5 -79.380641  
6 -79.380974  
7 -79.379276  
8 -79.381176  
9 -79.381949

In [42]:
## Venues with name Toronto in the neighbourhood and East York

In [39]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venuesT = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venuesT.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venuesT)

In [40]:
Toronto_venues = getNearbyVenues(names=df_ATC['Neighborhood'],
                                   latitudes=df_ATC['Latitude'],
                                   longitudes=df_ATC['Longitude']
                                  )

Davisville
Roselawn
North Toronto West
Forest Hill North & West
The Annex, North Midtown, Yorkville
Moore Park, Summerhill East
Lawrence Park
Davisville North
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
St. James Town
St. James Town, Cabbagetown
Kensington Market, Chinatown, Grange Park
University of Toronto, Harbord
Christie
Harbourfront East, Union Station, Toronto Islands
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
First Canadian Place, Underground city
Enclave of M5E
Studio District
India Bazaar, The Beaches West
The Danforth West, Riverdale
The Beaches
Enclave of M4L
Parkdale, Roncesvalles
High Park, The Junction South
Brockton, Parkdale Village, Exhibition Place
Little Portugal, Trinity
Duff

In [41]:
print(Toronto_venues.shape)
Toronto_venues.head()

(1481, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0   Davisville              43.704324               -79.38879   
1   Davisville              43.704324               -79.38879   
2   Davisville              43.704324               -79.38879   
3   Davisville              43.704324               -79.38879   
4   Davisville              43.704324               -79.38879   

                                 Venue  Venue Latitude  Venue Longitude  \
0                Jules Cafe Patisserie       43.704138       -79.388413   
1               Marigold Indian Bistro       43.702881       -79.388008   
2                            Zee Grill       43.704985       -79.388476   
3  Thobors Boulangerie Patisserie Café       43.704514       -79.388616   
4                          Sakae Sushi       43.704944       -79.388704   

       Venue Category  
0        Dessert Shop  
1   Indian Restaurant  
2  Seafood Restaurant  
3                Café  
4    Sushi Restaurant

In [42]:
Toronto_venues.groupby('Venue Category').count()

Neighborhood  Neighborhood Latitude  \
Venue Category                                                         
Adult Boutique                              1                      1   
Afghan Restaurant                           1                      1   
Airport                                     1                      1   
Airport Food Court                          1                      1   
Airport Gate                                1                      1   
Airport Lounge                              1                      1   
Airport Service                             3                      3   
Airport Terminal                            2                      2   
American Restaurant                        13                     13   
Antique Shop                                2                      2   
Aquarium                                    3                      3   
Art Gallery                                12                     12   
Arts & Crafts Store                         3                      3   
Asian Restaurant                           17                     17   
Auto Workshop                               1                      1   
BBQ Joint                                   2                      2   
Baby Store                                  1                      1   
Bagel Shop                                  7                      7   
Bakery                                     32                     32   
Bank                                       30                     30   
Bar                                        19                     19   
Baseball Stadium                            1                      1   
Basketball Stadium                          4                      4   
Beach                                       1                      1   
Bed & Breakfast                             1                      1   
Beer Bar                                   16                     16   
Beer Store                                  3                      3   
Belgian Restaurant                          2                      2   
Bistro                                      5                      5   
Board Shop                                  1                      1   
Boat or Ferry                               1                      1   
Bookstore                                   4                      4   
Boutique                                    4                      4   
Brazilian Restaurant                        2                      2   
Breakfast Spot                             10                     10   
Brewery                                     8                      8   
Bubble Tea Shop                             5                      5   
Burger Joint                               11                     11   
Burrito Place                              13                     13   
Bus Line                                    1                      1   
Butcher                                     2                      2   
Café                                       85                     85   
Cajun / Creole Restaurant                   1                      1   
Candy Store                                 2                      2   
Caribbean Restaurant                        3                      3   
Cheese Shop                                 5                      5   
Chinese Restaurant                          9                      9   
Chocolate Shop                              2                      2   
Church                                      1                      1   
Climbing Gym                                1                      1   
Clothing Store                             21                     21   
Cocktail Bar                               20                     20   
Coffee Shop                               157                    157   
College Arts Building                       1                      1   
C

In [43]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 216 uniques categories.


In [44]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head(20)

Yoga Studio  Adult Boutique  Afghan Restaurant  Airport  \
0             0               0                  0        0   
1             0               0                  0        0   
2             0               0                  0        0   
3             0               0                  0        0   
4             0               0                  0        0   
5             0               0                  0        0   
6             0               0                  0        0   
7             0               0                  0        0   
8             0               0                  0        0   
9             0               0                  0        0   
10            0               0                  0        0   
11            0               0                  0        0   
12            0               0                  0        0   
13            0               0                  0        0   
14            0               0                  0        0   
15            0               0                  0        0   
16            0               0                  0        0   
17            0               0                  0        0   
18            0               0                  0        0   
19            0               0                  0        0   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                    0             0               0                0   
1                    0             0               0                0   
2                    0             0               0                0   
3                    0             0               0                0   
4                    0             0               0                0   
5                    0             0               0                0   
6                    0             0               0                0   
7                    0             0               0                0   
8                    0             0               0                0   
9                    0             0               0                0   
10                   0             0               0                0   
11                   0             0               0                0   
12                   0             0               0                0   
13                   0             0               0                0   
14                   0             0               0                0   
15                   0             0               0                0   
16                   0             0               0                0   
17                   0             0               0                0   
18                   0             0               0                0   
19                   0             0               0                0   

    Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                  0                    0             0         0   
1                  0                    0             0         0   
2                  0                    0             0         0   
3                  0                    0             0         0   
4                  0                    0             0         0   
5                  0                    0             0         0   
6                  0                    0             0         0   
7                  0                    0             0         0   
8                  0                    0             0         0   
9                  0                    0             0         0   
10                 0                    0             0         0   
11                 0                    0             0         0   
12                 0                    0             0         0   
13                 0                    0             0         0   
14                 0                    0             0         0   
15                 0                    0 

In [45]:
Toronto_onehot.shape

(1481, 216)

In [46]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000   
2   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
3                                  Central Bay Street     0.015873   
4                                            Christie     0.000000   
5                                Church and Wellesley     0.014706   
6                      Commerce Court, Victoria Hotel     0.000000   
7                                          Davisville     0.000000   
8                                    Davisville North     0.000000   
9                        Dufferin, Dovercourt Village     0.000000   
10                                     Enclave of M4L     0.000000   
11                                     Enclave of M5E     0.000000   
12             First Canadian Place, Underground city     0.000000   
13                           Forest Hill North & West     0.000000   
14                           Garden District, Ryerson     0.000000   
15  Harbourfront East, Union Station, Toronto Islands     0.000000   
16                      High Park, The Junction South     0.000000   
17                     India Bazaar, The Beaches West     0.000000   
18          Kensington Market, Chinatown, Grange Park     0.000000   
19                                      Lawrence Park     0.000000   
20                           Little Portugal, Trinity     0.023810   
21                                 North Toronto West     0.000000   
22                             Parkdale, Roncesvalles     0.000000   
23                          Regent Park, Harbourfront     0.000000   
24                           Richmond, Adelaide, King     0.000000   
25                                           Rosedale     0.000000   
26                                           Roselawn     0.000000   
27                                 Runnymede, Swansea     0.032258   
28                                     St. James Town     0.000000   
29                        St. James Town, Cabbagetown     0.000000   
30                                    Studio District     0.000000   
31  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
32                The Annex, North Midtown, Yorkville     0.000000   
33                                        The Beaches     0.000000   
34                       The Danforth West, Riverdale     0.023256   
35           Toronto Dominion Centre, Design Exchange     0.000000   
36                     University of Toronto, Harbord     0.000000   

    Adult Boutique  Afghan Restaurant   Airport  Airport Food Court  \
0         0.000000           0.000000  0.000000            0.000000   
1         0.000000           0.000000  0.000000            0.000000   
2         0.000000           0.000000  0.071429            0.071429   
3         0.000000           0.000000  0.000000            0.000000   
4         0.000000           0.000000  0.000000            0.000000   
5         0.014706           0.014706  0.000000            0.000000   
6         0.000000           0.000000  0.000000            0.000000   
7         0.000000           0.000000  0.000000            0.000000   
8         0.000000           0.000000  0.000000            0.000000   
9         0.000000           0.000000  0.000000            0.000000   
10        0.000000           0.000000  0.000000            0.000000   
11        0.000000           0.000000  0.000000            0.000000   
12        0.000000           0.000000  0.000000            0.000000   
13        0.000000           0.000000  0.000000            0.000000   
14        0.000000           0.000000  0.000000            0.000000   
15        0.000000           0.000000  0.000000            0.000000   
16        0.000000           0.000000  0.000000            0.000000   
17        0.000000           0.000000  0.000000            0.000000   
18        0.000000           0.

In [47]:
Toronto_grouped.shape

(37, 216)

In [48]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
            venue  freq
0    Cocktail Bar  0.09
1     Coffee Shop  0.06
2  Sandwich Place  0.06
3          Bakery  0.06
4        Beer Bar  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0  Breakfast Spot  0.08
1  Sandwich Place  0.08
2          Bakery  0.08
3     Coffee Shop  0.08
4            Café  0.08


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0     Airport Service  0.21
1    Airport Terminal  0.14
2     Harbor / Marina  0.07
3             Airport  0.07
4  Airport Food Court  0.07


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.16
1       Sandwich Place  0.08
2     Sushi Restaurant  0.06
3  Japanese Restaurant  0.05
4                 Café  0.05


----Christie----
           venue  freq
0  Grocery Store  0.29
1           Café  0.21
2           Park  0.14
3     Baby Store  0.07
4      Nightclu

In [49]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [50]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park          Cocktail Bar   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery        Sandwich Place           Coffee Shop   
1        Sandwich Place                Bakery           Coffee Shop   
2      Airport Terminal      Sculpture Garden        Airport Lounge   
3        Sandwich Place      Sushi Restaurant                  Café   
4                  Café                  Park            Baby Store   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0    Seafood Restaurant        Farmers Market              Beer Bar   
1        Breakfast Spot                   Bar    Italian Restaurant   
2                 Plane   Rental Car Location       Harbor / Marina   
3    Italian Restaurant   Japanese Restaurant           Pizza Place   
4    Italian Restaurant             Nightclub            Restaurant   

           8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Vegetarian / Vegan Restaurant          Liquor Store                   Park  
1            Japanese Restaurant          Intersection              Nightclub  
2                   Airport Gate    Airport Food Court                Airport  
3                    Salad Place          Burger Joint                   Bank  
4                    Coffee Shop       Airport Service       Department Store

## Cluster Toronto Neighbourhoods

In [51]:
# number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 0, 1, 1, 1, 1, 1], dtype=int32)

In [53]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = df_ATC


Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood', how ='inner')

Toronto_merged.head()

PostalCode          Borough                         Neighborhood  \
47        M4S  Central Toronto                           Davisville   
63        M5N  Central Toronto                             Roselawn   
46        M4R  Central Toronto                   North Toronto West   
64        M5P  Central Toronto             Forest Hill North & West   
65        M5R  Central Toronto  The Annex, North Midtown, Yorkville   

     Latitude  Longitude  Cluster Labels    1st Most Common Venue  \
47  43.704324 -79.388790               1              Coffee Shop   
63  43.711695 -79.416936               4  Health & Beauty Service   
46  43.715383 -79.405678               1              Coffee Shop   
64  43.696948 -79.411307               3         Sushi Restaurant   
65  43.672710 -79.405678               1           Sandwich Place   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
47        Sandwich Place           Pizza Place          Dessert Shop   
63          Home Service                Garden             Wine Shop   
46        Clothing Store                  Café                 Diner   
64                  Park         Jewelry Store                 Trail   
65                  Café           Coffee Shop        History Museum   

   5th Most Common Venue        6th Most Common Venue  \
47      Sushi Restaurant           Seafood Restaurant   
63          Dance Studio  Eastern European Restaurant   
46            Restaurant                          Spa   
64             Wine Shop                   Donut Shop   
65          Liquor Store            Indian Restaurant   

        7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
47               Dance Studio    Italian Restaurant                  Café   
63                 Donut Shop      Doner Restaurant               Dog Run   
46       Fast Food Restaurant   Sporting Goods Shop            Bagel Shop   
64           Doner Restaurant               Dog Run   Distribution Center   
65  Middle Eastern Restaurant                   Pub             BBQ Joint   

   10th Most Common Venue  
47            Flower Shop  
63    Distribution Center  
46         Ice Cream Shop  
64         Discount Store  
65                   Park

In [54]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## Cluster 1, Neighbourhood, Christie (red cluster)
#### Christie appeears to be suburban with a grocery store as most popular venue, follwed by Cafe, Park and Baby Store

In [55]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
75  Downtown Toronto               0         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
75                  Café                  Park            Baby Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
75    Italian Restaurant             Nightclub            Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
75           Coffee Shop       Airport Service       Department Store

## Cluster 2 (purple Cluster) almost every neighbourhood in Toronto City
#### places were people go out and mingle, mainly restaurants and coffee shops for the top 6 or 7 venues in each neighbourhood

In [56]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
47         Central Toronto               1           Coffee Shop   
46         Central Toronto               1           Coffee Shop   
65         Central Toronto               1        Sandwich Place   
45         Central Toronto               1  Gym / Fitness Center   
49         Central Toronto               1           Coffee Shop   
52        Downtown Toronto               1   Japanese Restaurant   
53        Downtown Toronto               1           Coffee Shop   
54        Downtown Toronto               1           Coffee Shop   
56        Downtown Toronto               1          Cocktail Bar   
57        Downtown Toronto               1           Coffee Shop   
58        Downtown Toronto               1           Coffee Shop   
60        Downtown Toronto               1           Coffee Shop   
61        Downtown Toronto               1           Coffee Shop   
55        Downtown Toronto               1           Coffee Shop   
51        Downtown Toronto               1           Coffee Shop   
67        Downtown Toronto               1                  Café   
66        Downtown Toronto               1                  Café   
59        Downtown Toronto               1           Coffee Shop   
68        Downtown Toronto               1       Airport Service   
70        Downtown Toronto               1           Coffee Shop   
69  Downtown Toronto Stn A               1        Sandwich Place   
43            East Toronto               1           Coffee Shop   
42            East Toronto               1  Fast Food Restaurant   
41            East Toronto               1      Greek Restaurant   
37            East Toronto               1                   Pub   
87   East Toronto Business               1    Light Rail Station   
83            West Toronto               1           Coffee Shop   
82            West Toronto               1                  Café   
78            West Toronto               1                  Café   
77            West Toronto               1                   Bar   
76            West Toronto               1           Supermarket   
84            West Toronto               1           Coffee Shop   

            2nd Most Common Venue          3rd Most Common Venue  \
47                 Sandwich Place                    Pizza Place   
46                 Clothing Store                           Café   
65                           Café                    Coffee Shop   
45                     Playground                 Sandwich Place   
49                    Pizza Place                           Bank   
52               Sushi Restaurant                     Restaurant   
53                           Park                            Pub   
54                 Clothing Store                 Sandwich Place   
56                         Bakery                 Sandwich Place   
57                 Sandwich Place               Sushi Restaurant   
58                           Café                 Sandwich Place   
60                           Café                 Sandwich Place   
61                 Sandwich Place                           Café   
55                           Café             Italian Restaurant   
51                         Bakery                     Restaurant   
67  Vegetarian / Vegan Restaurant                    Coffee Shop   
66                    Coffee Shop                 Sandwich Place   
59                          Hotel                           Café   
68               Airport Terminal               Sculpture Garden   
70                           Café                 Sandwich Place   
69                    Coffee Shop                   Cocktail Bar   
43                      Gastropub                           Café   
42                 Sandwich Place                            Pub   
41             Italian Restaurant                    Coffee Shop   
37               Asian Restaurant                           Park   
87    

## Cluster 3 Lawerence Park (blue cluster)
#### recreation most common venue a Park with swim school College Gym and Dance Studio

In [57]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
44  Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
44              Bus Line           Swim School           College Gym   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
44          Dance Studio  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
44      Doner Restaurant               Dog Run    Distribution Center

## Cluster 4  Forest Hill and North West, and Rosedale (green cluster) in Downtown Toronto
#### By top venues combination of recreation and restaurants-Food

In [58]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
64   Central Toronto               3      Sushi Restaurant   
50  Downtown Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
64                  Park         Jewelry Store                 Trail   
50                 Trail            Playground          Cupcake Shop   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
64                    Wine Shop            Donut Shop      Doner Restaurant   
50  Eastern European Restaurant            Donut Shop      Doner Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
64               Dog Run   Distribution Center         Discount Store  
50               Dog Run   Distribution Center         Discount Store

## Cluster 5 Roselawn (Orange Cluster)
#### Top venues appear to be for mature people

In [59]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels    1st Most Common Venue  \
63  Central Toronto               4  Health & Beauty Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
63          Home Service                Garden             Wine Shop   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
63          Dance Studio  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
63      Doner Restaurant               Dog Run    Distribution Center

In [60]:
df_ATC

PostalCode                 Borough  \
47        M4S         Central Toronto   
63        M5N         Central Toronto   
46        M4R         Central Toronto   
64        M5P         Central Toronto   
65        M5R         Central Toronto   
48        M4T         Central Toronto   
44        M4N         Central Toronto   
45        M4P         Central Toronto   
49        M4V         Central Toronto   
50        M4W        Downtown Toronto   
52        M4Y        Downtown Toronto   
53        M5A        Downtown Toronto   
54        M5B        Downtown Toronto   
56        M5E        Downtown Toronto   
57        M5G        Downtown Toronto   
58        M5H        Downtown Toronto   
60        M5K        Downtown Toronto   
61        M5L        Downtown Toronto   
55        M5C        Downtown Toronto   
51        M4X        Downtown Toronto   
67        M5T        Downtown Toronto   
66        M5S        Downtown Toronto   
75        M6G        Downtown Toronto   
59        M5J        Downtown Toronto   
68        M5V        Downtown Toronto   
70        M5X        Downtown Toronto   
69        M5W  Downtown Toronto Stn A   
43        M4M            East Toronto   
42        M4L            East Toronto   
41        M4K            East Toronto   
37        M4E            East Toronto   
87        M7Y   East Toronto Business   
83        M6R            West Toronto   
82        M6P            West Toronto   
78        M6K            West Toronto   
77        M6J            West Toronto   
76        M6H            West Toronto   
84        M6S            West Toronto   

                                         Neighborhood   Latitude  Longitude  
47                                         Davisville  43.704324 -79.388790  
63                                           Roselawn  43.711695 -79.416936  
46                                 North Toronto West  43.715383 -79.405678  
64                           Forest Hill North & West  43.696948 -79.411307  
65                The Annex, North Midtown, Yorkville  43.672710 -79.405678  
48                        Moore Park, Summerhill East  43.689574 -79.383160  
44                                      Lawrence Park  43.728020 -79.388790  
45                                   Davisville North  43.712751 -79.390197  
49  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049  
50                                           Rosedale  43.679563 -79.377529  
52                               Church and Wellesley  43.665860 -79.383160  
53                          Regent Park, Harbourfront  43.654260 -79.360636  
54                           Garden District, Ryerson  43.657162 -79.378937  
56                                        Berczy Park  43.644771 -79.373306  
57                                 Central Bay Street  43.657952 -79.387383  
58                           Richmond, Adelaide, King  43.650571 -79.384568  
60           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
61                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
55                                     St. James Town  43.651494 -79.375418  
51                        St. James Town, Cabbagetown  43.667967 -79.367675  
67          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
66                     University of Toronto, Harbord  43.662696 -79.400049  
75                                           Christie  43.669542 -79.422564  
59  Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
68  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
70             First Canadian Place, Underground city  43.648429 -79.382280  
69                                     Enclave of M5E  43.646435 -79.374846  
43                                    Studio District  43.659526 -79.340923  
42                     India Bazaar, The Beaches West  43.668999 -79.315572  
41                       The Danforth West, Riverdale  43.67955